olive/systems/azureml/aml_system.py

```diff
@@ -373,6 +378,36 @@ def pass_runner_pipeline():
 
         return pass_runner_pipeline()
 
+    def _create_data_script_inputs_and_args(self, data_configs: Optional[Dict[str, Any]] = None):
```

Comment:
just enable this feature so to change so huge files seems be a little bit hacky.

Are there any examples show the usage? Is there any better way to achieve similar functionality?

Reply:
This could be a quick design doc but I can summarize here:

Use case: some data configs may have `user_script` and `script_dir`, and some passes will use those data configs, like LoRA pass and some quantization passes. When running these passes on AML, only the pass config was submitted to the AML compute, with local `user_script` and `script_dir`, we need to upload all data configs user scripts and script dirs to AML, and in the meantime update the `user_script` and `script_dir` with AML file path in pass config.

So the challenges here are:
1. The data configs are various with different names and different user scripts and script dirs. Each data config's user script and script dir need to be uploaded separately to AML. So we will need to parse each user script and each script dir separately.
2. The data config was nested in the pass config, and only the pass config was submitted to the AML. The `run_pass` method for system has no knowledge about data configs. When creating AML job inputs and args for data config, we can easily get all non-None `user_script` and `script_dir` in data configs from olive workflow and add them to AML job, while when parsing back in aml pass runner, since data configs are nested in the pass config, we can hardly know which parser arguments are needed for this pass.
3. Then, we need to update data config in pass config with new AML file path.

To tackle challenge 1:
- We need data configs section now at system level (unless we collect all nested data configs from pass config, which is not preferrable).
- Since each `user_script` and each `script_dir` have to be uploaded to AML separately, we need a naming convention for each input, which can easily differentiate user script and script dir for each data config, and can be easily parsed back in aml system runner. So I decide to use `data_name_user_script` and `data_name_script_dir` for the input names.

To tackle challenge 2, there are a couple of options:
1. When submitting the job to AML, upload data configs as a file to AML job also. Go through each data config and figure out which data config has user script and script dir, then we can know the parser argument for all user scripts and script dirs.
Pros: Easy check on AML pass runner.
Cons: 1. A new file will be uploaded to aml. 2. DataConfig class now is not serializable. Need to support serialization for DataConfig class, which means we need to support serialization for all classes related to DataConfig class.

2. All checks are done in aml pass runner. Add a new  `ParamCategory.DATA_CONFIG` to each data config pass (do not reply on the param name like ending in `data_config`). Like how we do for [pass path now](https://github.com/microsoft/Olive/blob/main/olive/systems/azureml/aml_pass_runner.py#L43C23-L43C23), check if the pass has this `DATA_CONFIG` param and if this param config has `user_script` or `script_dir`.
Pros: No need change for DataConfig class. 
Cons: Tedious parsing logic in aml pass runner.

Above, I prefer the second approach.

To tackle challenge 3:
With the option 1 for challenge 2, even we can get all data configs param arguments for the pass, we still need to go over each pass parameter to find the correct data config parameter. But if we go with option 2 for challenge 2, we can update the `user_script` and `script_dir` aml path directly when we found it. So I still go for option 2.

Reply:
I agree with the solutions for challenge 2/3 but not for challenge 1. 

I don't think it is necessary to keep a system or engine level information for `data_configs`. There are multiple issues with it:
1. We cannot guarantee that the workflow config's `data_configs` member has all possible data config objects inside it. A pass config can also directly specify the data_config with referencing to the workflows `data_configs`. 
2. We are unnecessarily uploading all data configs for all passes even though they might not use them at all. 
3. data name is not a good key to use. We cannot guarantee that the user will not provide an incompatible name. This PR shows that we had to update the example to make it work.

I think it is better to do the same thing we do for `path_params`:
1. We already know what the data config params are.
2. There is not much nesting. It is just one level below the pass config so is the same as iterating over the pass config. This is already done in the pass runner and when you iterate over the `data_configs` in this PR. 
3. instead of using data_name as key, we can just use the pass param name. It is guaranteed to be unique for a pass since we would only use the data configs involved with the pass.

Reply:
for challenge 1, is it possible to have one consolidated user_script and script_dir for all the data?

Reply:
@jambayk I thought about the cases you mentioned, but I decide to make our life easier, like maybe we should only allow the data config in data configs section, and have a naming rule for data configs. The reason I'm using data name as key rather param name is, I'm not sure if there will be any case that different data config params will use the same data config. If so, duplicated files and folders will be uploaded to AML. So regarding to uploading unnecessary data configs, I was also thinking doing a cross validation between data configs and the data configs in pass configs. But I'm open to this if we don't have any use case that same data config is used for different data config param in the pass.

Reply:
@guotuofeng I don't think that would work. Our current logic treat user_script and script_dir differently. We still need to differentiate them in AML pass runner if so.

Reply:
we could refactor our code to make it support this behavior. I suppose

Reply:
Can you explain more about "consolidated user_script and script_dir for all the data"?

Reply:
@jambayk another option could be, iterating over the pass config to find the data configs that are used for this pass, but still keep data config name as param name to avoid duplicating. With this, we do not need data config arg for system. This would be better.

Reply:
"We cannot guarantee that the workflow config's data_configs member has all possible data config objects inside it. "
It'd be fine to put this restriction and have pass specific data_config only refer the data config from the data_configs.

Reply:
> Can you explain more about "consolidated user_script and script_dir for all the data"?

we can add an assumption that all user written code need to be put into a single user_script for Olive. If there are multiple user_scripts file, the user need merge them into a single one to use Olive. 

This assumption should not a big deal for Olive users.

Reply:
@xiaoyu-work I am fine with your initial suggestion and devang's about data configs only being defined in the `data_configs` section. 

But that also seems like an extra restriction since we would need to iterate through the pass config anyway as you mentioned in 
> another option could be, iterating over the pass config to find the data configs that are used for this pass, but still keep data config name as param name to avoid duplicating. With this, we do not need data config arg for system. This would be better.
I agree this sounds like a better compromise since your original concern was about duplicated upload. We will however need to add some validations for the names of the data config so that can be used as command line arguments and also to make sure there are no duplicate names.

Reply:
@guotuofeng I think the work to have a single user script for a single workflow has greater scope than this PR and might be better dealt separately, preferably before this PR? This also came up in a previous discussion with @trajepl 

We have user scripts in many different parts of olive: models, metrics, pass configs, data configs, etc. We would need to come up with a way to provide the user script or module to the different components if they don't keep track of their own user scripts. We would also need to do this while also keeping the modular nature of the components. What do you think about this @devang-ml?

Reply:
Yes, you are correct. the single user script is another work item unrelated to the PR. I am ok with current approach to upload multiple user script way. 

However, we need not introduce the change at system level.